[doc] Refactor type system (#5984)
taichi-dev/taichi: docs/lang/articles/type_system/type.md

# Type system

> Taichi is a statically typed programming language. The type of a variable in the Taichi scope is determined at compile time. Once a variable is declared, you cannot bind its name to a value of a different type afterward.

Reviewer: ```suggestion
Taichi is a statically typed programming language. The type of a variable in the Taichi scope is determined at compile time. After a variable is declared, you cannot assign to it a value of a different type.
```

I don't really understand what "bind its name to a value of a different type" means :-)

Let's see a quick example:

```python
@ti.kernel
def test():
    x = 1  # x is the integer 1
    x = 3.14  # 3.14 is cast to 3 hence x is equal to 3
    x = ti.Vector([1, 1])  # Error!
```

Line 3 indicates that `x` should be an integer since it is assigned by an integer value 1 upon its declaration. In Line 4, we re-assign a floating-point number 3.14 to `x`. This scalar will be automatically cast to integer 3 to match the type of `x`; hence, `x` takes the value 3 after this line. Line 5 will raise a syntax error when we try to assign a `ti.Vector` to `x` because `ti.Vector` is a different data type, which cannot be cast into an integer.

We can briefly summarize the data types supported by Taichi as follows:

Taichi supports two categories of data types in the [Taichi scope](../kernels/syntax.md#taichi-scope): primitive types and compound types.

> - Primitive types: Various commonly used numerical data types, such as `int32`, `uint8`, and `float64`.

Reviewer: ```suggestion
- Primitive types: Various commonly used numerical data types, such as `ti.i32 (int32)`, `ti.u8 (uint8)`, and `ti.f64 (float64)`.
```

> - Compound types: Array-like or struct-like data types, including `matrix`, `struct`, and `ndarray`, which comprise multiple members of primitive types or other compound types.

Reviewer: ```suggestion
- Compound types: Array-like or struct-like data types, including `ti.types.matrix`, `ti.types.struct`, and `ti.types.ndarray`, which comprise multiple members of primitive types or other compound types.
```

# Type system

Taichi is a statically typed programming language. The type of a variable in the Taichi scope is determined at compile time. After a variable is declared, you cannot assign to it a value of a different type.

Let's see a quick example:

```python
@ti.kernel
def test():
    x = 1  # x is the integer 1
    x = 3.14  # 3.14 is casted to 3; hence, x takes the value 3
    x = ti.Vector([1, 1])  # Error!
```

Line 3 indicates that `x` should be an integer since it is assigned by an integer value 1 upon its declaration. In Line 4, we re-assign a floating-point number 3.14 to `x`. This scalar will be automatically cast to integer 3 to match the type of `x`; hence, `x` takes the value 3 after this line. Line 5 will raise a syntax error when we try to assign a `ti.Vector` to `x` because `ti.Vector` is a different data type, which cannot be cast into an integer.

We can briefly summarize the data types supported by Taichi as follows:

Taichi supports two categories of data types in the [Taichi scope](../kernels/syntax.md#taichi-scope): primitive types and compound types.

> - Primitive types: Various commonly used numerical data types, such as `ti.i32 (int32)`, `ti.u8 (uint8)`, and `ti.f64 (float64)`.

Reviewer: ```suggestion
- Primitive types: Various commonly used numerical data types, such as `ti.i32` (`int32`), `ti.u8` (`uint8`), and `ti.f64` (`float64`).
```

> :o:: Requiring extensions for the backend.

> ### Customize the default primitive types for integers and floating-point numbers

Reviewer: Let's leave a note on **numeric literals such as `42`, `3.14` will have default integral/floating-point types**.

When you call the `ti.init()` method to initialize the Taichi runtime, Taichi automatically sets the default types for integer and floating-point arithmetic:

- The default integer type in Taichi is `ti.i32`.
- The default floating-point type in Taichi is `ti.f32`.

> You can customize these default primitive types by explicitly specifying them in the `init()` call:

Reviewer 2: ```suggestion
You can customize the default setting by explicitly specifying the primitive data type(s) you want to use in the `init()` call:
```

> :o:: Requiring extensions for the backend.

### Customize the default primitive types for integers and floating-point numbers

When you call the `ti.init()` method to initialize the Taichi runtime, Taichi automatically applies the default primitive data types:

- The default integer type in Taichi is `ti.i32`.
- The default floating-point type in Taichi is `ti.f32`.

You can customize the default setting by explicitly specifying the primitive data type(s) you want to use in the `init()` call:

```python
ti.init(default_ip=ti.i64)  # Set the default integer type to ti.i64
ti.init(default_fp=ti.f64)  # Set the default floating-point type to ti.f64
```

> It should be noted that the numeric literals in the Taichi scope also have default integer/floating-point types. For example, if the default floating-point type is `ti.f32`, then a numeric literal `3.14159265358979` in the Taichi scope will be cast into a 32-bit floating-point number, hence only accurate up to about seven decimal places.

Reviewer: ```suggestion
It should be noted that the numeric literals in the Taichi scope also have default integer/floating-point types. For example, if the default floating-point type is `ti.f32`, then a numeric literal `3.14159265358979` in the Taichi scope will be cast into a 32-bit floating-point number, hence only accurate up to about seven decimal places. If you work on high-precision workloads such as numeric simulations for engineering, use `ti.f64` as the `default_fp`.
```

> It should be noted that the numeric literals in the Taichi scope also have default integer/floating-point types. For example, if the default floating-point type is `ti.f32`, then a numeric literal `3.14159265358979` in the Taichi scope will be cast into a 32-bit floating-point number, hence only accurate up to about seven decimal places. If you work on high-precision workloads such as numeric simulations for engineering, use `ti.f64` as the `default_fp`.

Reviewer 2: ```suggestion
It should be noted that the numeric literals in the Taichi scope also have default integer/floating-point types. For example, if the default floating-point type is `ti.f32`, then a numeric literal `3.14159265358979` in the Taichi scope will be cast into a 32-bit floating-point number, hence a precision of up to about seven decimal digits. If you work on high-precision workloads, such as numeric simulations for engineering, use `ti.f64` as the `default_fp`.
```

      a = 3.14
      x = int(a)    # 3
      y = float(a)  # 3.14
      z = ti.i32(a)  # 3
      w = ti.f64(a)  # 3.14
  ```

### Implicit typecasting

Implicit typecasting occurs when you *accidentally* put or assign a value in a place where a different data type is expected.

:::caution WARNING
As a rule of thumb, implicit typecasting is a major source of bugs. And Taichi does *not* recommend resorting to this mechanism.

:::

> Implicit typecasting can happen in binary operations or in assignments, as explained below.

Reviewer: Typecast has a different meaning from `type cast`.

>       y1 = ti.f64(a)  # 3.14

Reviewer: Can we rename `x1`, `y1` to `z` and `w`?

      print(a)  # 1.0
  ```

> - Example 2: Variable `a` is initialized with type `int` and immediately reassigned `3.14`. The reassignment implicitly casts `3.14` from `float` to `int` with a warning because the type of `a` is `int` and has a low precision than `float`:

Reviewer: ```suggestion
- Example 2: Variable `a` is initialized with type `int` and immediately reassigned `3.14`. The reassignment implicitly casts `3.14` from `float` to `int` with a warning because the type of `a` is `int` and has a lower precision than `float`:
```

## Compound types

Compound types are user-defined data types, which comprise multiple members. Supported compound types include vectors, matrices, ndarrays, and structs.

Taichi allows you to use all types supplied in the `ti.types` module as scaffolds to customize *higher-level* compound types.

:::note

The `ndarray` type is discussed in another article [interacting with External Arrays](../basic/external.md).

:::

### Matrix and Vector types

You can use the two functions `ti.types.matrix()` and `ti.types.vector()` to create your own matrix and vector types:

```python
vec4d = ti.types.vector(4, ti.f64)  # 64-bit floating-point 4D vector type
mat4x3i = ti.types.matrix(4, 3, int)  # integer 4x3 matrix type
```

> Such types can be directly called to create the corresponding vector/matrix instances, and can also be used as type hints in function arguments and struct members. For example:

Reviewer 2: ```suggestion
You can directly call the customized vector/matrix type to instantiate vectors/matrices or use them as type hints in function arguments and struct members. For example:
```

```python

> v = vec4d(1, 2, 3, 4)  # Create a vector instance, here v = [1.0, 2.0 3.0 4.0]

Reviewer: ```suggestion
v = vec4d(1, 2, 3, 4)  # Create a vector instance, here v = [1.0 2.0 3.0 4.0]
```

@ti.func
def length(w: vec4d):  # vec4d as type hint
    return w.norm()

@ti.kernel
def test():
    print(length(v))
```

> 

Reviewer: We need one more subsection here: "Relationship between ti.Vector/Matrix and ti.types.vector/matrix"

### Struct types and dataclass

> You can use the funtion `ti.types.struct()` to create a struct type, and use this type as a template to create struct instances. For example suppose you are using Taichi to represent a sphere. A sphere in the 3D space can be abstracted with its center and radius. In the following example, you call `ti.types.vector()` and `ti.types.struct()` to create compound types `vec3` and `sphere_type`. These two types are the *higher-level* compound types that fit better with your scenario. Once you have customized your compound types, you can use them as templates to create two instances of spheres (initialize two local variables `sphere1` and `sphere2`):

Reviewer 2: ```suggestion
You can use the function` ti.types.struct()` to create a struct type, and use this type as a template to create struct instances. For example, suppose you want to represent a sphere in Taichi. A 3D sphere can be abstracted with its center and radius. You can call `ti.types.vector()` and `ti.types.struct()` to create compound types `vec3` and `sphere_type`, respectively. These two types are the *higher-level* compound types that fit better with your scenario. Then, you can use them as templates to create two instances of spheres (initialize two local variables `sphere1` and `sphere2`):
```

```python
vec4d = ti.types.vector(4, ti.f64)  # a 64-bit floating-point 4D vector type
mat4x3i = ti.types.matrix(4, 3, int)  # a 4x3 integer matrix type
```

You can use the customized types to instantiate vectors and matrices or annotate the data types of function arguments and struct members. For example:

```python
v = vec4d(1, 2, 3, 4)  # Create a vector instance, here v = [1.0 2.0 3.0 4.0]

@ti.func
def length(w: vec4d):  # vec4d as type hint
    return w.norm()

@ti.kernel
def test():
    print(length(v))
```

### Relationship between `ti.Matrix/ti.Vector` and `ti.types.matrix/ti.types.vector`

In Taichi vectors are all matrices, except that they have only one column (hence are column vectors). The function call `ti.Vector()` simply converts the input into a matrix of a single column:

```python
v = ti.Vector([1, 1, 1]) # equivalent to v = ti.Matrix([[1], [1], [1]])
```

Similarily, the type returned by the function call `ti.types.vector()` is nothing more than a matrix type except it always has only one column:

```
v = ti.types.vector(3, float)  # equivalent to v = ti.types.matrix(3, 1, float)

> ```

Reviewer 2: Delete

You can use the two functions `ti.types.matrix()` and `ti.types.vector()` to create your own matrix and vector types:

```python
vec4d = ti.types.vector(4, ti.f64)  # a 64-bit floating-point 4D vector type
mat4x3i = ti.types.matrix(4, 3, int)  # a 4x3 integer matrix type
```

You can use these customized types to instantiate vectors and matrices or annotate the data types of function arguments and struct members. For example:

```python
v = vec4d(1, 2, 3, 4)  # Create a vector instance, here v = [1.0 2.0 3.0 4.0]

@ti.func
def length(w: vec4d):  # vec4d as type hint
    return w.norm()

@ti.kernel
def test():
    print(length(v))
```

In practice, `ti.types.matrix` only would suffice your need for vector/matrix customization because Taichi treats vectors as a special kind of matrices, i.e., matrices with one column.

In fact, calling the function `ti.types.vector()` produces a matrix type of a single column:

```
v = ti.types.vector(3, float)  # equivalent to v = ti.types.matrix(3, 1, float)
```

Similarly, `ti.Vector()` simply converts the input into a matrix of a single column:

```python
v = ti.Vector([1, 1, 1])  # equivalent to v = ti.Matrix([[1], [1], [1]])

> ```

Reviewer: I don't the two `v`s are the same. The former is a type, while the latter is a value.

sphere2 = sphere_type(center=vec3([1, 1, 1]), radius=1.0)
```

> When there are a dozen of members in a struct, the above way of using `ti.types.struct` may make your code look messy, and there is a better way to do so. Indeed, we recommend you use the decorator `@ti.dataclass` as a more intuitive way to define your custom struct type in the form of a class. The dataclass is merely a thin wrapper of struct:

Reviewer 2: ```suggestion
When a struct contains too many members, `ti.types.struct` may make your code look messy. Taichi offers a more intuitive way to define a struct: The decorator `@ti.dataclass` is a thin wrapper that allows you to customize the struct type in the form of a class. 
```

sphere2 = sphere_type(center=vec3([1, 1, 1]), radius=1.0)
```

When a struct contains many members, `ti.types.struct` may make your code look messy. Taichi offers a more intuitive way to define a struct: The decorator `@ti.dataclass` is a thin wrapper of the struct type:

```python
@ti.dataclass
class Sphere:
    center: vec3
    radius: float
```

The code above serves the same purpose as the line below does but provides better readability:

```python
Sphere = ti.types.struct(center=vec3, radius=float)
```

> Another advantage of using `@ti.dataclass` over `ti.types.struct` is that you can define member functions in a dataclass and call them in the Taichi scope, make them very handy for OOP programming. See the article [objective data-oriented programming](../advanced/odop.md) for more detail.

Reviewer 2: ```suggestion
Another advantage of using `@ti.dataclass` over `ti.types.struct` is that you can define member functions in a dataclass and call them in the Taichi scope, making object-oriented programming (OOP) possible. See the article [objective data-oriented programming](../advanced/odop.md) for more details.
```
